rust/dialog-storage/src/storage/backend/s3.rs

```diff
@@ -0,0 +1,2482 @@
+    pub fn new(mut url: Url, prefix: Option<&str>, continuation_token: Option<&str>) -> Self {
+        url.query_pairs_mut().append_pair("list-type", "2");
+        if let Some(prefix) = prefix {
+            url.query_pairs_mut().append_pair("prefix", prefix);
+        }
+        if let Some(token) = continuation_token {
+            url.query_pairs_mut()
+                .append_pair("continuation-token", token);
+        }
+        Self { url }
+    }
+}
+
+impl Invocation for List {
+    fn method(&self) -> &'static str {
+        "GET"
+    }
+
+    fn url(&self) -> &Url {
+        &self.url
+    }
+}
+impl Request for List {}
+
+/// Response from S3 ListObjectsV2 API (simplified).
+#[derive(Debug)]
+pub struct ListResult {
+    /// Object keys returned in this response.
+    pub keys: Vec<String>,
+    /// If true, there are more results to fetch.
+    pub is_truncated: bool,
+    /// Token to use for fetching the next page of results.
+    pub next_continuation_token: Option<String>,
+}
+
+/// S3-safe key encoding that preserves path structure.
+///
+/// Keys are treated as `/`-delimited paths. Each path component is checked:
+/// - If it contains only safe characters (alphanumeric, `-`, `_`, `.`), it's kept as-is
+/// - Otherwise, it's base58-encoded and prefixed with `!`
+///
+/// The `!` character is used as a prefix marker because it's in AWS S3's
+/// "safe for use" list and unlikely to appear at the start of path components.
+///
+/// Examples:
+/// - `remote/main` → `remote/main` (all components safe)
+/// - `remote/user@example` → `remote/!<base58>` (@ is unsafe, encode component)
+/// - `foo/bar/baz` → `foo/bar/baz` (all safe)
+pub fn encode_s3_key(bytes: &[u8]) -> String {
+    use base58::ToBase58;
+
+    let key_str = String::from_utf8_lossy(bytes);
+    let components: Vec<&str> = key_str.split('/').collect();
+
+    let encoded_components: Vec<String> = components
+        .iter()
+        .map(|component| {
+            // Check if component contains only safe characters
+            let is_safe = component.bytes().all(|b| {
+                matches!(b,
+                    b'a'..=b'z' | b'A'..=b'Z' | b'0'..=b'9' | b'-' | b'_' | b'.'
+                )
+            });
+
+            if is_safe && !component.is_empty() {
+                component.to_string()
+            } else {
+                // Base58 encode and prefix with !
+                format!("!{}", component.as_bytes().to_base58())
+            }
+        })
+        .collect();
+
+    encoded_components.join("/")
+}
+
+/// Decode an S3-encoded key back to bytes.
+///
+/// Path components starting with `!` are base58-decoded.
+/// Other components are used as-is.
+pub fn decode_s3_key(encoded: &str) -> Result<Vec<u8>, S3StorageError> {
+    use base58::FromBase58;
+
+    let components: Vec<&str> = encoded.split('/').collect();
+    let mut decoded_components: Vec<Vec<u8>> = Vec::new();
+
+    for component in components {
```

Comment:
You should be able to collect into a `Result` type to simplify this (and also note that we avoid the second `Vec` allocation).

```rs
encoded.split('/').map(|component| {
  if let Some(encoded_part) = component.strip_prefix('!') {
    encoded_part.from_base58().map_err(|e| {
      S3StorageError::SerializationFailed(format!(
          "Invalid base58 encoding in component '{}': {:?}",
          component, e
      ))
    })
  } else {
    Ok(component.as_bytes().to_vec())
  }
}).collect::<Result<Vec<u8>, S3StorageError>>()
```